src/main/java/graphql/nadel/engine/OverallQueryTransformer.java

```diff
@@ -221,62 +229,66 @@ QueryTransformationResult transformMergedFields(
                         serviceContext,
                         removedFieldMap
                 );
-                // Case happens when the high level field is removed
-                if (newField == null) {
-                    newField = field.transform(builder -> builder.selectionSet(SelectionSet.newSelectionSet().build()));
-                }
-                // if all child fields of the high level field are removed then the top-level field is nulled
+                return newFieldCF.thenApply(newField -> {
+                    // Case happens when the high level field is removed
+                    if (newField == null) {
+                        newField = field.transform(builder -> builder.selectionSet(SelectionSet.newSelectionSet().build()));
+                    }
+                    // if all child fields of the high level field are removed then the top-level field is nulled
+                    GraphQLOutputType fieldType = rootType.getFieldDefinition(field.getName()).getType();
+                    newField = ArtificialFieldUtils.maybeAddUnderscoreTypeName(nadelContext, newField, fieldType);
+                    return newField;
 
-                GraphQLOutputType fieldType = rootType.getFieldDefinition(field.getName()).getType();
-                newField = ArtificialFieldUtils.maybeAddUnderscoreTypeName(nadelContext, newField, fieldType);
-                return newField;
+                });
             });
-            transformedFields.addAll(transformed);
-            MergedField transformedMergedField = MergedField.newMergedField(transformed).build();
-            transformedMergedFields.add(transformedMergedField);
-
+            transformedFieldsCF.addAll(transformedCF);
```

Comment:
Idk if there's a `flatMap` function? Should be used here if there is one. It would read better